src/Compilers/Core/Portable/CommandLine/AnalyzerConfig.cs

```diff
@@ -0,0 +1,485 @@
+            foreach (var config in analyzerConfigs)
+            {
+                // Create an array of regexes with each entry corresponding to the same index
+                // in <see cref="EditorConfig.NamedSections"/>.
+                var builder = ArrayBuilder<Regex>.GetInstance(config.NamedSections.Length);
+                foreach (var section in config.NamedSections)
+                {
+                    string regex = AnalyzerConfig.TryCompileSectionNameToRegEx(section.Name);
+                    builder.Add(new Regex(regex, RegexOptions.Compiled));
+                }
+
+                Debug.Assert(builder.Count == config.NamedSections.Length);
+
+                allRegexes.Add(config, builder.ToImmutableAndFree());
+            }
+
+            Debug.Assert(allRegexes.Count == analyzerConfigs.Count);
+
+            var treeOptionsBuilder = ImmutableDictionary.CreateBuilder<string, ReportDiagnostic>(
+                CaseInsensitiveComparison.Comparer);
+            var analyzerOptionsBuilder = ImmutableDictionary.CreateBuilder<string, string>(
+                CaseInsensitiveComparison.Comparer);
+            foreach (var sourceFile in sourcePaths)
+            {
+                var normalizedPath = pathWithForwardSlashSeparators(sourceFile);
+                // The editorconfig paths are sorted from shortest to longest, so matches
+                // are resolved from most nested to least nested, where last setting wins
+                foreach (AnalyzerConfig config in analyzerConfigs)
+                {
+                    if (normalizedPath.StartsWith(config.NormalizedDirectory, StringComparison.Ordinal))
+                    {
+                        int dirLength = config.NormalizedDirectory.Length;
+                        // Leave '/' if the normalized directory ends with a '/'. This can happen if
+                        // we're in a root directory (e.g. '/' or 'Z:/'). The section matching
+                        // always expects that the relative path start with a '/'. 
+                        if (config.NormalizedDirectory[dirLength - 1] == '/')
+                        {
+                            dirLength--;
+                        }
+                        string relativePath = normalizedPath.Substring(dirLength);
+
+                        ImmutableArray<Regex> regexes = allRegexes[config];
+                        for (int sectionIndex = 0; sectionIndex < regexes.Length; sectionIndex++)
+                        {
+                            if (regexes[sectionIndex].IsMatch(relativePath))
+                            {
+                                var section = config.NamedSections[sectionIndex];
+                                addOptions(section, treeOptionsBuilder, analyzerOptionsBuilder, config.PathToFile);
+                            }
+                        }
+                    }
+                }
+
+                // Avoid extra allocations by using null.
```

Comment:
>Avoid extra allocations by using null. [](start = 19, length = 38)

`ImmutableDictionary<,>.Empty` is a singleton.

Reply:
But `ImmutableDictionary.Create<string, ReportDiagnostic>(CaseInsensitiveComparison.Comparer)` is not. I do have a singleton on SyntaxTree for this, but the downstream API maps null to this singleton, so I'm not sure I see the value in making the singleton internal instead of protected.